src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+
+			hasJobs = false;
+			break;
+		}
+
+		/* we load the database name and username here as we are still in a transaction */
+		char *databaseName = get_database_name(MyDatabaseId);
+		char *userName = GetUserNameFromId(task->owner, false);
+
+		MemoryContextSwitchTo(oldContext);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		/*
+		 * The background worker needs to be started outside of the transaction, otherwise
+		 * it will complain about leaking shared memory segments used, among other things,
+		 * to communicate the output of the backend.
+		 */
+		dsm_segment *seg = NULL;
+		BackgroundWorkerHandle *handle =
+			StartCitusBackgroundJobExecuter(databaseName, userName, task->command,
+											task->taskid, &seg);
+
+		if (handle == NULL)
+		{
+			ereport(ERROR, (errmsg("unable to start background worker for background "
+								   "task execution")));
+		}
+
+		pid_t pid = 0;
+		GetBackgroundWorkerPid(handle, &pid);
+
+		ereport(LOG, (errmsg("found task with jobid/taskid: %ld/%ld",
+							 task->jobid, task->taskid)));
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		task->status = BACKGROUND_TASK_STATUS_RUNNING;
+		SET_NULLABLE_FIELD(task, pid, pid);
+
+		/* Update task status to indicate it is running */
+		UpdateBackgroundTask(task);
+		UpdateBackgroundJob(task->jobid);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		bool hadError = false;
+		bool cancelled = false;
+
+		/*
+		 * We reset the old message (if present). This will only retain the last message
+		 * in the catalog. Otherwise it would concatenate all retries.
+		 */
+		if (task->message)
+		{
+			pfree(task->message);
+		}
+		task->message = NULL;
+
+		while (GetBackgroundWorkerPid(handle, &pid) != BGWH_STOPPED)
+		{
+			int latchFlags = WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH;
+			int rc = WaitLatch(MyLatch, latchFlags, (long) 1000, PG_WAIT_EXTENSION);
+
+			/* emergency bailout if postmaster has died */
+			if (rc & WL_POSTMASTER_DEATH)
+			{
+				proc_exit(1);
+			}
+
+			if (rc & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError, &cancelled);
+
+			shm_mq_detach(responseq);
+		}
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		{
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError, &cancelled);
```

Comment:
how reliable is the "cancelled" flag?

Reply:
I see what you are saying. If we cancel a backemd, but it consumes the signal and responds with a different signal back we might miss the cancel status here.

That would require a bit of thought on how to handle concurrent updates to the Task data. Because we would need to re-read it from the catalog and understand what we want to change and what might have changed since we last read the Task.

Reply:
We don't rely on the cancelled tag anymore, and instead we re-read the Task from the catalog after the executor exits. We rely on the state described in the updated metadata to understand if we were `cancelling` a task.

This also introduces an actual cancelling state between a `running` task and a `cancelled` task due to the delay between the job being cancelled and the actual output of the task being written to the catalog.

There was a test that actually deduced if a running task was being able to be cancelled by checking for the output message.